### PR TITLE
[mkldnn-v1.0] Add MKL-DNN sum concat

### PR DESCRIPTION
**Pull request by rongzha1**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here
@ciyongch @TaoLv @ZhennanQin @PatricZhao @wuxun-zhang

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@zixuanweeei please take a review?

File: src/operator/nn/mkldnn/mkldnn_concat.cc [hunk 1]
```diff
@@ -20,62 +20,45 @@
 /*!
  * \file mkldnn_concat.cc
  * \brief
```
Comment (zixuanweeei) [inline src/operator/nn/mkldnn/mkldnn_concat.cc]: nit: A short brief may be preferred 😄

File: src/operator/nn/mkldnn/mkldnn_sum.cc [hunk 2]
```diff
@@ -150,18 +119,22 @@ void MKLDNNSumForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
...
-                                                   fwd.fwd_pd.dst_primitive_desc(),
+                                                   fwd.fwd_pd.dst_desc(),
                                                    req,
                                                    &in_bufs[0]);
-  fwd.SetNewMem(data_mem, *out_mem.second);
-  MKLDNNStream::Get()->RegisterPrim(fwd.GetFwd());
+  mkldnn_args_map_t net_args;
+  net_args.insert({MKLDNN_ARG_DST, *out_mem.second});
+  for (int i = 0; i < num_inputs; ++i) {
+    net_args.insert({MKLDNN_ARG_MULTIPLE_SRC + i, *data_mem[i]});
```
Comment (zixuanweeei) [inline src/operator/nn/mkldnn/mkldnn_sum.cc]: @ZhennanQin has recommended a better `insert` approach, see https://github.com/apache/incubator-mxnet/pull/16199#discussion_r326038204.

File: src/operator/nn/mkldnn/mkldnn_sum.cc [hunk 3]
```diff
@@ -54,41 +54,33 @@ void MKLDNNSum(const mkldnn::memory &arr1,
     in_mem2 = tmp_memory2;
   }
   mkldnn::sum::primitive_desc sum_pd(output_pd, scales, input_pds, CpuEngine::Get()->get_engine());
-  std::unordered_map<int, mkldnn::memory> args = {
+  mkldnn_args_map_t args = {
     { MKLDNN_ARG_MULTIPLE_SRC, *in_mem1 },
     { MKLDNN_ARG_MULTIPLE_SRC + 1, *in_mem2 },
     { MKLDNN_ARG_DST, out },
   };
```
Comment (zixuanweeei) [inline src/operator/nn/mkldnn/mkldnn_sum.cc]: Also needs to avoid temporary pairs here?

File: src/operator/nn/mkldnn/mkldnn_concat.cc (hunk 1 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_concat.cc]: @ciyongch Can you help to add some brief ?

File: src/operator/nn/mkldnn/mkldnn_sum.cc (hunk 2 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_sum.cc]: OK  thanks

File: src/operator/nn/mkldnn/mkldnn_concat.cc (hunk 1 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_concat.cc]: seems just file head style, no mkldnn file add brief head

File: src/operator/nn/mkldnn/mkldnn_sum.cc (hunk 3 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_sum.cc]: I'm not sure this less efficient than serveral assignment.

**Comment by TaoLv** [comment]
@perdasilva @marcoabreu Could you please take a look at the CI failure? Looks like it's related to the CI environment?

```
+ python tests/python-pytest/onnx/backend_test.py

Traceback (most recent call last):

  File "tests/python-pytest/onnx/backend_test.py", line 89, in <module>

    unittest.TextTestRunner().run(test_suite(BACKEND_TESTS.enable_report()))

  File "/usr/local/lib/python2.7/dist-packages/onnx/backend/test/runner/__init__.py", line 96, in enable_report

    import pytest  # type: ignore

  File "/usr/local/lib/python2.7/dist-packages/pytest.py", line 13, in <module>

    from _pytest.fixtures import fixture, yield_fixture

  File "/usr/local/lib/python2.7/dist-packages/_pytest/fixtures.py", line 832, in <module>

    class FixtureFunctionMarker(object):

  File "/usr/local/lib/python2.7/dist-packages/_pytest/fixtures.py", line 834, in FixtureFunctionMarker

    params = attr.ib(convert=attr.converters.optional(tuple))

TypeError: attrib() got an unexpected keyword argument 'convert'
```

**Comment by marcoabreu** [comment]
My first guess is that possibly pytest got updated in our Docker containers

**Comment by TaoLv** [comment]
@rongzha1, Pooling is merged. Please rebase this PR. Thanks.

**Comment by lanking520** [comment]
About the ONNX test, the dummy PR seemed working: https://github.com/apache/incubator-mxnet/issues/16394. Please rebase with master and try again

**Comment by pengzhao-intel** [comment]
@rongzha1 please rebase the code and re-run before merging.

**Comment by pengzhao-intel** [comment]
merging now.

**Pull request merged**